Automattic/wp-calypso
# Plans (Storage): Enable storage selection for current plan, and other fixes/refactors

Fixes https://github.com/Automattic/wp-calypso/issues/90905 (start of the pit)
Fixes https://github.com/Automattic/wp-calypso/issues/90947
Fixes https://github.com/Automattic/wp-calypso/issues/90949
Fixes https://github.com/Automattic/wp-calypso/issues/91047
Based on https://github.com/Automattic/wp-calypso/pull/90989

## Proposed Changes

- Enables showing the storage dropdown selector for the current plan in spotlight
- Renders the combined storage volume for the site/plan when a storage add-on has been purchased
- Renders the combined storage volume in the dropdown selections (as a fallback in case of similar regressions to https://github.com/Automattic/wp-calypso/issues/91047)
- Some refactors to clean up the code a bit, start moving things into their own folder (there are enough pieces to storage logic that merit its own little space in the package). We will follow-up with more refactors down the line, consolidate and centralise more of the logic, avoid repeating things across Comparison & Features Grid, etc.

### Media

<img width="600" alt="Screenshot 2024-05-30 at 10 15 44 AM" src="https://github.com/Automattic/wp-calypso/assets/1705499/cb901fea-fa91-4428-ac30-9abb607685fc">


<img width="600" alt="Screenshot 2024-05-30 at 10 11 59 AM" src="https://github.com/Automattic/wp-calypso/assets/1705499/3567d900-cca3-4ea5-92ee-c077c61ad151">

<img width="600" alt="Screenshot 2024-05-30 at 10 13 12 AM" src="https://github.com/Automattic/wp-calypso/assets/1705499/d8d518a7-5681-43c7-bfce-c94f57ca7e20">


### Needs Clarity

The remaining parts (to my mind) that need clarity:

1. I am going by the assumption that we show the updated/full storage for the site under the current plan after a purchase upgrade. This should align with what we want to do with the new Spotlight designs: p9Jlb4-bSl-p2
2. Should the price of the plans in the grid when a storage upgrade has been purchased reflect the new overall cost to the user? Before a purchase, we update the yearly prices in the grid to reflect the additional cost of extra storage - should this not be the case after the purchase? e.g.:

**On current plan:**

with the 100GB storage shown, should the price be `70.96` and not `25`?

<img width="600" alt="Screenshot 2024-05-23 at 3 07 01 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/3603f00d-05fe-47f7-8b2b-e4dedd40bb69">


**On upgrade plan (Entrepreneur):**

with the 100GB storage shown, should the price be `90.96` and not `45`?

<img width="600" alt="Screenshot 2024-05-23 at 3 11 08 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/3d3c4114-2b5b-4268-a91c-347cdfc3cabf">


## Why are these changes being made?


Addresses all of https://github.com/Automattic/wp-calypso/issues/90905, https://github.com/Automattic/wp-calypso/issues/90947, https://github.com/Automattic/wp-calypso/issues/90949, https://github.com/Automattic/wp-calypso/issues/91047

## Testing Instructions



- Go to `/start/plans` and confirm storage selection and rendering works as before
- Create a site and go to `/plans/[site]` and confirm the same
- Confirm the current/spotlight plan allows for selecting/purchasing additional storage
- Purchase 50GB additional storage and revisit `/plans/[site with extra storage]`
    - Confirm the media above: 
        - the total storage reflected in the storage label rendered for plans that allow storage upgrades
        - storage selection no longer enabled
        - (optional - the fallback) remove the `storageAddOn.purchased` constraints from the code and reload. The selectors for current and Entrepreneur plans should show selection that reflects what the site would have upgraded to, as outlined in https://github.com/Automattic/wp-calypso/issues/91047 (so 150GB and 200GB, instead of 100GB and 150GB)

## Pre-merge Checklist
